client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,248 @@
+		if let Some(shared) = self.shared.upgrade() {
+			shared.lock().unsubscribe(self.subs_id);
+		}
+	}
+}
+
+impl<M, R> Hub<M, R> {
+	/// Create a new instance of Hub (with default value for the Registry).
+	pub fn new(tracing_key: &'static str) -> Self
+	where
+		R: Default,
+	{
+		Self::new_with_registry(tracing_key, Default::default())
+	}
+
+	/// Create a new instance of Hub over the initialized Registry.
+	pub fn new_with_registry(tracing_key: &'static str, registry: R) -> Self {
+		let shared =
+			Shared { registry, sinks: Default::default(), id_sequence: Default::default() };
+		let shared = Arc::new(Mutex::new(shared));
+		Self { tracing_key, shared }
+	}
+
+	/// Subscribe to this Hub using the `subs_key: K`.
+	///
+	/// A subscription with a key `K` is possible if the Registry implements `Subscribe<K>`.
+	pub fn subscribe<K>(&self, subs_key: K) -> Receiver<M, R>
+	where
+		R: Subscribe<K> + Unsubscribe,
+	{
+		let mut shared = self.shared.lock();
+
+		let subs_id = shared.id_sequence.next_id();
+		let (tx, rx) = crate::mpsc::tracing_unbounded(self.tracing_key);
+		assert!(shared.sinks.insert(subs_id, tx).is_none(), "Used IDSequence to create another ID. Should be unique until u64 is overflowed. Should be unique.");
+		shared.registry.subscribe(subs_key, subs_id);
+
+		Receiver { shared: Arc::downgrade(&self.shared), subs_id, rx }
+	}
+
+	/// Send the message produced with `Trigger`.
+	///
+	/// This is possible if the registry implements `Dispatch<Trigger, Item = M>`.
+	pub fn send<Trigger>(&self, trigger: Trigger)
+	where
+		R: Dispatch<Trigger, Item = M>,
+	{
+		let mut shared = self.shared.lock();
+		let (registry, sinks) = shared.get_mut();
+
+		registry.dispatch(trigger, |subs_id, item| {
+			if let Some(tx) = sinks.get_mut(&subs_id) {
+				if let Err(send_err) = tx.unbounded_send(item) {
+					log::warn!("Sink with SubsID = {} failed to perform unbounded_send: {} ({} as Dispatch<{}, Item = {}>::dispatch(...))", subs_id, send_err, std::any::type_name::<R>(),
+					std::any::type_name::<Trigger>(),
+					std::any::type_name::<M>());
+				}
+			} else {
+				log::warn!(
+					"No Sink for SubsID = {} ({} as Dispatch<{}, Item = {}>::dispatch(...))",
+					subs_id,
+					std::any::type_name::<R>(),
+					std::any::type_name::<Trigger>(),
+					std::any::type_name::<M>(),
+				);
+			}
+		});
+	}
+}
+
+impl<M, R> Shared<M, R> {
+	fn get_mut(&mut self) -> (&mut R, &mut HashMap<SeqID, TracingUnboundedSender<M>>) {
+		(&mut self.registry, &mut self.sinks)
+	}
+
+	fn unsubscribe(&mut self, subs_id: SeqID)
+	where
+		R: Unsubscribe,
+	{
+		self.registry.unsubscribe(subs_id);
+		self.sinks.remove(&subs_id);
+	}
+}
+
+impl<M, R> Clone for Hub<M, R> {
+	fn clone(&self) -> Self {
+		Self { tracing_key: self.tracing_key, shared: self.shared.clone() }
+	}
+}
+
+impl<M, R> Unpin for Receiver<M, R> where R: Unsubscribe {}
+
+impl<M, R> Stream for Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	type Item = <TracingUnboundedReceiver<M> as Stream>::Item;
```

Comment:
Such way of specifying the type is everywhere in this repository (e.g. [here](https://github.com/paritytech/substrate/blob/dc4cf6c2c73ef5af9197b3e73fb590b1a045d93e/client/network/src/protocol.rs#L1377) ).

My logic was: since the `Stream` implementation for the `Receiver<M,R>` just forwards the logic to the underlying channel, it would only be consistent to forward the type definitions as well.

Reply:
But we know that `M` is the message here. It makes no sense when you directly know the type, to do some extra gymnastics.